variable/test/comparison_test.cpp

```diff
@@ -70,3 +70,105 @@ TEST(ComparisonTest, tolerance_type_mismatch) {
   EXPECT_THROW(is_approx(a, b, 0 /*implicit int*/), except::TypeError);
   EXPECT_NO_THROW(is_approx(a, b, 0.0f));
 }
+
+TEST(ComparisonTest, less_variances_test) {
+  const auto a = makeVariable<float>(Dims{Dim::X}, Shape{2}, Values{1.0, 2.0},
+                                     Variances{1.0, 1.0});
+  const auto b = makeVariable<float>(Dims{Dim::X}, Shape{2}, Values{0.0, 3.0},
+                                     Variances{1.0, 1.0});
+  EXPECT_THROW(less(a, b), std::runtime_error);
+}
+
+TEST(ComparisonTest, less_dtypes_test) {
+  const auto a = makeVariable<float>(Dims{Dim::X}, Shape{2}, Values{1.0, 2.0});
+  const auto b = makeVariable<int>(Dims{Dim::X}, Shape{2}, Values{0, 1});
+  EXPECT_THROW(less(a, b), std::runtime_error);
+}
+
+TEST(ComparisonTest, less_units_test) {
+  const auto a = makeVariable<double>(Dims{Dim::X}, Shape{2}, Values{1.0, 2.0});
+  auto b = makeVariable<double>(Dims{Dim::X}, Shape{2}, Values{0.0, 3.0});
+  b.setUnit(units::m);
+  EXPECT_THROW(less(a, b), std::runtime_error);
+}
+
+template <typename T> class LessTest : public ::testing::Test {};
+template <typename T> class GreaterTest : public ::testing::Test {};
+template <typename T> class LessEqualTest : public ::testing::Test {};
+template <typename T> class GreaterEqualTest : public ::testing::Test {};
+template <typename T> class EqualTest : public ::testing::Test {};
+template <typename T> class NotEqualTest : public ::testing::Test {};
+
+using CompareTestTypes = ::testing::Types<double, float, int64_t, int32_t>;
+TYPED_TEST_SUITE(GreaterTest, CompareTestTypes);
+TYPED_TEST_SUITE(LessTest, CompareTestTypes);
+TYPED_TEST_SUITE(LessEqualTest, CompareTestTypes);
+TYPED_TEST_SUITE(GreaterEqualTest, CompareTestTypes);
+TYPED_TEST_SUITE(EqualTest, CompareTestTypes);
+TYPED_TEST_SUITE(NotEqualTest, CompareTestTypes);
+
+TYPED_TEST(LessTest, value) {
+  using T = TypeParam;
+  T y = 1;
+  T x = 2;
+  const auto a = makeVariable<T>(Dims{Dim::X}, Shape{1}, Values{x});
```

Comment:
Note that there is a shorter way to make a 0-D variable: Just multiple a number by a  `Unit`, e.g., `1.0 * units::Unit(units::m)` (this sould actually be reduced to `1.0 * units::m` after a pending cleanup task). Maybe this would make the test code marginally more readable (if also the typed test would be dropped, since this is retesting, see above):

```cpp
EXPECT_EQ(less(1.0 * units::Unit(units::m), 1.1 * units::Unit(units::m)), true * units::Unit(units::dimensionless));
```

Another option for shorter and more readable tests (avoiding a lot of repitition in this file) on this level would be to combine everything in one test fixture (for all comparison operators), defining:

```cpp
const auto a = 1.0 * units::Unit(units::m);
const auto b = 2.0 * units::Unit(units::m);
const auto true_ = 1.0 * units::Unit(units::dimensionless);
const auto false_ = 2.0 * units::Unit(units::dimensionless);
```
Then add test cases for each of the operators, e.g.,
```cpp
EXPECT_EQ(less(a, b), true_);
EXPECT_EQ(less(b, a), false_);
EXPECT_EQ(less(a, a), false_);
```
and so on.